cloudbuild.yaml

```diff
@@ -143,23 +177,105 @@ steps:
         -var=gcs_bucket=gke-aieco-jupyter-$SHORT_SHA \
         -auto-approve -no-color
 
-        cd /workspace/applications/ray/
+    allowFailure: true
+    waitFor: ['test jupyterhub']
+
+  - id: 'test rag'
+    name: 'gcr.io/$PROJECT_ID/terraform'
+    entrypoint: 'sh'
+    args:
+      - '-c'
+      - |
+        set -e
+
+        # Get kube config
+        gcloud container clusters get-credentials \
+        ml-$SHORT_SHA-$_PR_NUMBER-cluster \
+        --location $_REGION \
+        --project $PROJECT_ID 
+
+        cd /workspace/modules/jupyter/tests
+        python3 change_jupyter_config.py
+
+        cd /workspace/applications/rag/
+        terraform init
+        terraform apply \
+        -var-file=workloads.tfvars \
+        -var=project_id=$PROJECT_ID \
+        -var=cluster_name=ml-$SHORT_SHA-$_PR_NUMBER-cluster \
+        -var=kubernetes_namespace=rag-$SHORT_SHA \
+        -var=gcs_bucket=gke-aieco-rag-$SHORT_SHA \
+        -auto-approve -no-color -lock=false
+        echo "pass" > /workspace/rag_tf_result.txt
+
+        # Validate Ray: Make sure pods are running
+        kubectl wait --all pods -n rag-$SHORT_SHA --for=condition=Ready --timeout=300s
+        kubectl port-forward -n rag-$SHORT_SHA service/example-cluster-kuberay-head-svc 8265:8265 &
+        # Wait port-forwarding to take its place
+        sleep 5s
+
+        # Validate Ray: Check dashboard
+        ray job submit --working-dir ./tests \
+        --address=http://127.0.0.1:8265 -- python ray_job.py 
+        echo "pass" > /workspace/rag_ray_dashboard_result.txt
+
+        # Validate Jupyterhub: Get hub url
+        kubectl get services -n rag-$SHORT_SHA
+        kubectl get service proxy-public -n rag-$SHORT_SHA --output jsonpath='{.status.loadBalancer.ingress[0].ip}' > /workspace/rag_jupyterhub_host_url.txt
+        echo "HOST URL is " $(cat /workspace/rag_jupyterhub_host_url.txt)
+
+        # Validate Jupyterhub: Test Hub
+        cd /workspace/modules/jupyter/tests
+        python3 test_hub.py $(cat /workspace/rag_jupyterhub_host_url.txt)
+        echo "pass" > /workspace/rag_jupyterhub_test_result.txt
+
+        # Validate RAG: Test rag frontend
+        kubectl port-forward -n rag-$SHORT_SHA service/rag-frontend 8081:8080 &
+        # Wait port-forwarding to take its place
+        sleep 5s
+
+        cd /workspace/applications/rag/tests
+        python3 test_frontend.py "127.0.0.1:8081"
+        echo "pass" > /workspace/rag_frontend_result.txt
+
+    allowFailure: true
+    waitFor: ['cleanup jupyterhub']
```

Comment:
wondering if this can run in parallel to jupyter run, any specific reason we wait for `cleanup jupyterhub` here?

Reply:
This has beed fixed as per the comment